### PR TITLE
Enable LTO for GCC Builds

### PR DESCRIPTION
**Pull request by chenhao94**

### CODE REVIEW

**Comment by chenhao94** [comment]
@xkszltl As you see there will be a bunch of "undefined references" when enabling LTO for GCC:

https://github.com/cyfitech/cris-core/actions/runs/4218280339/jobs/7322668113#step:5:329

Any ideas on why and how to resolve this?

**Comment by xkszltl** [comment]
I've seen this before elsewhere but cannot recall why.

```
ERROR: /__w/cris-core/cris-core/src/msg_recorder/tools/BUILD:5:15: Linking src/msg_recorder/tools/ldb_record_convert failed: (Exit 1): gcc-10-wrapper failed: error executing command /__w/cris-core/cris-core/run/toolchain/gcc-10-wrapper @bazel-out/k8-opt/bin/src/msg_recorder/tools/ldb_record_convert-2.params
```
Please check that .params locally first to see if the linked libs make sense.
If so, readelf/nm the lib to see if symbols are really there, with correct linkage.
You may need to use the gcc-blessed version (something like gcc-readelf-10) for LTO objects.

**Comment by chenhao94** [comment]
@zptan @xkszltl  Let's start from a non-controversial case with `-flto -fuse-linker-plugin`.

https://github.com/cyfitech/cris-core/actions/runs/4220501562/jobs/7326832188#step:5:678

```
ERROR: /__w/cris-core/cris-core/tests/BUILD:79:14: Linking tests/node_test failed: (Exit 1): gcc-10-wrapper failed: error executing command /__w/cris-core/cris-core/run/toolchain/gcc-10-wrapper @bazel-out/k8-opt/bin/tests/node_test-2.params

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
external/gtest/googlemock/src/gmock-spec-builders.cc:600: error: undefined reference to 'testing::FLAGS_gmock_catch_leaked_mocks'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::FLAGS_gmock_verbose[abi:cxx11]'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::FLAGS_gmock_verbose[abi:cxx11]'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::internal::g_gmock_implicit_sequence'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::internal::ThreadLocal<testing::Sequence*>::ThreadLocal()'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::internal::g_gmock_implicit_sequence'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'testing::internal::ThreadLocal<testing::Sequence*>::~ThreadLocal()'
/usr/include/c++/10/streambuf:205: error: undefined reference to 'std::map<void const*, testing::internal::CallReaction, std::less<void const*>, std::allocator<std::pair<void const* const, testing::internal::CallReaction> > >::~map()'
collect2: error: ld returned 1 exit status
```

Local `@bazel-out/k8-opt/bin/tests/node_test-2.params`:
```
> cat bazel-out/k8-opt/bin/tests/node_test-2.params                                             
-o  
bazel-out/k8-opt/bin/tests/node_test
-fuse-ld=gold
-Wl,-no-as-needed
-Wl,-z,relro,-z,now
-B/home/chenhao/cris-core/run/toolchain
-pass-exit-codes
-Wl,--gc-sections
bazel-out/k8-opt/bin/tests/_objs/node_test/node_test.o
-Wl,--start-lib
bazel-out/k8-opt/bin/_objs/msg/message.o
bazel-out/k8-opt/bin/_objs/msg/node.o
-Wl,--end-lib      
-Wl,--start-lib
bazel-out/k8-opt/bin/_objs/sched/job_lock_queue.o
bazel-out/k8-opt/bin/_objs/sched/job_lockfree_queue.o
bazel-out/k8-opt/bin/_objs/sched/job_queue.o
bazel-out/k8-opt/bin/_objs/sched/job_runner.o
bazel-out/k8-opt/bin/_objs/sched/job_runner_config_parser.o
bazel-out/k8-opt/bin/_objs/sched/spin_impl.o
bazel-out/k8-opt/bin/_objs/sched/spin_mutex.o
-Wl,--end-lib
bazel-out/k8-opt/bin/external/simdjson/libsimdjson/lib/libsimdjson.a
-Wl,--start-lib
bazel-out/k8-opt/bin/tests/_objs/cris_gtest_main/cris_gtest_main.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/_objs/signal/cr_signal.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/_objs/timer/timer.o
bazel-out/k8-opt/bin/_objs/timer/timer_disable.o
bazel-out/k8-opt/bin/_objs/timer/timer_enable.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/_objs/utils/defs.o
bazel-out/k8-opt/bin/_objs/utils/time.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/demangle.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/logging.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/raw_logging.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/signalhandler.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/symbolize.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/utilities.o
bazel-out/k8-opt/bin/external/com_github_google_glog/_objs/glog/vlog_is_on.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/external/com_github_gflags_gflags/_objs/gflags/gflags.o
bazel-out/k8-opt/bin/external/com_github_gflags_gflags/_objs/gflags/gflags_completions.o
bazel-out/k8-opt/bin/external/com_github_gflags_gflags/_objs/gflags/gflags_reporting.o
-Wl,--end-lib
-Wl,--start-lib
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock-cardinalities.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock-internal-utils.o 
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock-matchers.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock-spec-builders.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-death-test.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-filepath.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-port.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-printers.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-test-part.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest-typed-test.o
bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gtest.o
-Wl,--end-lib
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-ldl
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-Wall
-Werror
-Wextra
-Wno-fuse-ld-path
-lpthread
-pthread
-flto
-fuse-linker-plugin
-lstdc++
-lm
```

And I use `'testing::FLAGS_gmock_verbose[abi:cxx11]'` as example:

```
> nm --version && printf '\n\n' && nm -ACn bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock.o | grep 'FLAGS_gmock_verbose'
GNU nm (GNU Binutils for Ubuntu) 2.34
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) any later version.
This program has absolutely no warranty.


bazel-out/k8-opt/bin/external/gtest/_objs/gtest/gmock.o:00000000 T testing::FLAGS_gmock_verbose[abi:cxx11]
```
```

**Comment by xkszltl** [comment]
我记得好像start-lib/end-lib这种写法有些不支持的case

**Comment by xkszltl** [comment]
你要不要试试把node_test-2.params里的东西拿出来用一个长命令手动跑。
可以从后往前逐渐加lib，看看什么时候炸

**Comment by chenhao94** [comment]
Seems gold linker LTO does not work well with `--start/end-lib` flags. So we use the normal `.a` archives and LTO works just fine.

@xkszltl The remaining problem is, do we need to find the ThinLTO alternatives for GCC instead of the default LTO? If so then I can open an issue and assign it to both of us.

File: scripts/toolchain_gcc.sh [hunk 1]
```diff
@@ -10,7 +10,9 @@ if [ -d 'run' ] && [ -w 'run' ] || [ -w '.' ]; then
     cat << ________EOF | sed 's/^        //' > run/toolchain.bazelrc
         # GCC-specific Bazel Configurations
 
-        # TODO(chenhao94, xkszltl): Add LTO options
-        # build:rel --config="lto"
+        build:rel --config="lto"
+        build:lto --nostart_end_lib
+        build:lto --copt="-flto"
+        build:lto --linkopt="-flto"
```
[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain_gcc.sh:16]: 没必要，bazel对gcc会自动加这条

[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain_gcc.sh:16]: @zptan 你可能是环境有什么奇怪的地方或者cache不干净

[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain_gcc.sh:16]: bazel版本是多少，我记得bazel是不是在4到5的时候改过这个行为

**Comment by xkszltl** [comment]
> @xkszltl The remaining problem is, do we need to find the ThinLTO alternatives for GCC instead of the default LTO? If so then I can open an issue and assign it to both of us.

Yes if possible.

File: scripts/toolchain_gcc.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (xkszltl) [reply scripts/toolchain_gcc.sh:16]: 我也觉得应该手动写上，甚至于自动加这个行为都是UB吧？我反而挺在意如果我们写别的bazel还会不会自动加

**Review by xkszltl — APPROVED** [review]

File: scripts/toolchain_gcc.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain_gcc.sh:16]: > bazel版本是多少，我记得bazel是不是在4到5的时候改过这个行为

其实没改过，只是有一个版本找lld有问题所以我们之前手动加过

[the remark this answers is not in the document]
Comment (chenhao94) [reply scripts/toolchain_gcc.sh:18]: @zptan 更新了。现在这个写法你那边可以么

**Pull request merged**